src/Components/Endpoints/src/Rendering/SSRRenderModeBoundary.cs

```diff
@@ -106,15 +116,52 @@ private void Prerender(RenderTreeBuilder builder)
             var serverComponentSerializer = httpContext.RequestServices.GetRequiredService<ServerComponentSerializer>();
 
             var invocationId = EndpointHtmlRenderer.GetOrCreateInvocationId(httpContext);
-            serverMarker = serverComponentSerializer.SerializeInvocation(invocationId, _componentType, parameters, _prerender);
+            serverMarker = serverComponentSerializer.SerializeInvocation(invocationId, _componentType, parameters, _markerKey, _prerender);
         }
 
         WebAssemblyComponentMarker? webAssemblyMarker = null;
         if (_renderMode is WebAssemblyRenderMode or AutoRenderMode)
         {
-            webAssemblyMarker = WebAssemblyComponentSerializer.SerializeInvocation(_componentType, parameters, _prerender);
+            webAssemblyMarker = WebAssemblyComponentSerializer.SerializeInvocation(_componentType, parameters, _markerKey, _prerender);
         }
 
         return (serverMarker, webAssemblyMarker);
     }
+
+    private string GenerateMarkerKey(int sequence, object? key)
+    {
+        var componentTypeNameHash = GetOrComputeComponentTypeNameHash(_componentType);
+        return $"{componentTypeNameHash}:{sequence}:{(key as IFormattable)?.ToString(null, CultureInfo.InvariantCulture)}";
+    }
+
+    private static string GetOrComputeComponentTypeNameHash(Type componentType)
+    {
+        if (!s_componentTypeNameHashCache.TryGetValue(componentType, out var hash))
+        {
+            hash = ComputeComponentTypeNameHash(componentType);
+            s_componentTypeNameHashCache.Add(componentType, hash);
```

Comment:
Writes to a `Dictionary<T,U>` aren't threadsafe according to the [docs](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2?view=net-8.0#thread-safety) so should this be a `ConcurrentDictionary`? Or do you know that this particular TryGetValue+Add pattern would be threadsafe?